File: python/clean_dataset.py
import os
import matplotlib.pyplot as plt
import numpy as np
import csv
from pygam import LinearGAM, s
from tqdm import tqdm

def gam_smoothing(values, lam):
    x = np.arange(len(values)).reshape(-1, 1)
    y = np.array(values)
    
    if np.all(y == 0) or np.std(y) == 0:
        return y.tolist()

    gam = LinearGAM(s(0, lam=lam)).fit(x, y)
    y_pred = gam.predict(x)
    
    return y_pred.tolist()
    
def detect_outliers(data, window_size=5, threshold=0.3):
    """
    Detects and replaces outliers in a dataset using a sliding window approach.
    Parameters:
    data (list or array-like): The input dataset to be cleaned.
    window_size (int, optional): The size of the window to calculate the mean and standard deviation. Default is 5.
    threshold (float, optional): The threshold factor to determine if a data point is an outlier. Default is 0.3.
    Returns:
    list: A list containing the cleaned dataset with outliers replaced by the mean of the window.
    Example:
    >>> data = [1, 2, 1, 1, 100, 1, 2, 1, 1]
    >>> detect_outliers(data)
    [1, 2, 1, 1, 1.4, 1, 2, 1, 1]
    """
    cleaned_data = []
    for i in range(len(data)):
        start, end = max(0, i - window_size), min(len(data), i + window_size + 1)
        window = data[start:end]
        
        mean, std = np.mean(window), np.std(window)
        
        if abs(data[i] - mean) > threshold * std:
            cleaned_data.append(mean) 
        else:
            cleaned_data.append(data[i])

    return cleaned_data

smooth = 0.0005
def clean_dataset(pixel_dataset):
    grouped_data = {}
    for data in pixel_dataset:
        key = (data['x'], data['y'])
        if key not in grouped_data:
            grouped_data[key] = []
        grouped_data[key].append(data)

    progress_bar = tqdm(total=len(grouped_data), desc=f"Cleaning dataset")
    for key, data in grouped_data.items():
        ndre,ndmi,psri,ndvi = [],[], [], []
        for i in range(len(data)):
            ndre.append(float(data[i]['ndre']))
            ndmi.append(float(data[i]['ndmi']))
            psri.append(float(data[i]['psri']))
            ndvi.append(float(data[i]['ndvi']))


        ndmi = detect_outliers(ndmi)
        psri = detect_outliers(psri)
        ndre = detect_outliers(ndre)
        ndvi = detect_outliers(ndvi)

        ndvi = gam_smoothing(ndvi,smooth)
        ndmi = gam_smoothing(ndmi,smooth)
        psri = gam_smoothing(psri,smooth)
        ndre = gam_smoothing(ndre,smooth)
        
        valid_data = []
        for i in range(len(data)):
            ndmi_value = str(ndmi.pop(0))
            psri_value = str(psri.pop(0))
            ndre_value = str(ndre.pop(0))
            ndvi_value = str(ndvi.pop(0))
            if float(ndmi_value) == 0 or float(psri_value) == 0 or float(ndre_value) == 0 or float(ndvi_value) == 0:
                continue
            data[i]['ndmi'] = ndmi_value
            data[i]['psri'] = psri_value
            data[i]['ndre'] = ndre_value
            data[i]['ndvi'] = ndvi_value
            data[i]['precipitation'] = str(data[i]['precipitation'])
            data[i]['temperature'] = str(data[i]['temperature'])
            data[i]['humidity'] = str(data[i]['humidity'])
            valid_data.append(data[i])
        grouped_data[key] = valid_data
        progress_bar.update(1)

    new_array = []
    for key, data in grouped_data.items():
        for entry in data:
            new_array.append(entry)

    if len(new_array) > 0:
        return new_array 
    else:
        print(f"No valid data for")








if __name__ == "__main__":
    clean_dataset("data/not_clean", "boi_preto")

File: go/clean_dataset.go
package main

import (
	"fmt"
	"math"
	"sync"
)

func detectOutliers(data []float64, windowSize int, threshold float64) []float64 {
	cleanedData := make([]float64, len(data))

	for i := range data {
		start := int(math.Max(0, float64(i-windowSize)))
		end := int(math.Min(float64(len(data)), float64(i+windowSize+1)))
		window := data[start:end]

		mean, std := calculateMeanAndStd(window)

		if math.Abs(data[i]-mean) > threshold*std {
			cleanedData[i] = mean
		} else {
			cleanedData[i] = data[i]
		}
	}

	return cleanedData
}

func calculateMeanAndStd(data []float64) (mean, std float64) {
	sum := 0.0
	for _, v := range data {
		sum += v
	}
	mean = sum / float64(len(data))

	variance := 0.0
	for _, v := range data {
		variance += math.Pow(v-mean, 2)
	}
	std = math.Sqrt(variance / float64(len(data)))

	return mean, std
}

func gamSmoothing(values []float64, lam float64) []float64 {
	// Placeholder for GAM smoothing logic
	// Replace this with the actual implementation later
	return values
}

func cleanDataset(pixelDataset []PixelData) []PixelData {
	groupedData := make(map[[2]int][]PixelData)

	// Group data by (x, y)
	for _, data := range pixelDataset {
		key := [2]int{data.X, data.Y}
		groupedData[key] = append(groupedData[key], data)
	}

	var wg sync.WaitGroup
	mu := sync.Mutex{}
	newArray := []PixelData{}

	for key, data := range groupedData {
		wg.Add(1)
		go func(key [2]int, data []PixelData) {
			defer wg.Done()

			var ndre, ndmi, psri, ndvi []float64
			for _, d := range data {
				ndre = append(ndre, d.NDRE)
				ndmi = append(ndmi, d.NDMI)
				psri = append(psri, d.PSRI)
				ndvi = append(ndvi, d.NDVI)
			}

			ndmi = detectOutliers(ndmi, 5, 0.3)
			psri = detectOutliers(psri, 5, 0.3)
			ndre = detectOutliers(ndre, 5, 0.3)
			ndvi = detectOutliers(ndvi, 5, 0.3)

			ndvi = gamSmoothing(ndvi, 0.0005)
			ndmi = gamSmoothing(ndmi, 0.0005)
			psri = gamSmoothing(psri, 0.0005)
			ndre = gamSmoothing(ndre, 0.0005)

			validData := []PixelData{}
			for i := range data {
				if ndmi[i] == 0 || psri[i] == 0 || ndre[i] == 0 || ndvi[i] == 0 {
					continue
				}
				data[i].NDMI = ndmi[i]
				data[i].PSRI = psri[i]
				data[i].NDRE = ndre[i]
				data[i].NDVI = ndvi[i]
				validData = append(validData, data[i])
			}

			mu.Lock()
			newArray = append(newArray, validData...)
			mu.Unlock()
		}(key, data)
	}

	wg.Wait()

	if len(newArray) > 0 {
		return newArray
	} else {
		fmt.Println("No valid data found")
		return nil
	}
}
